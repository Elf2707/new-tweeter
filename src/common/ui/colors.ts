export default {
  $primaryColor: '#FF530D',
  $secondaryColor: '#E82C0C',
  $accentColor: '#FF0DFF',

  // Text palette
  $textColor: '#000',
  $subTextColor: '#555',

  // Light palette
  $white: '#FFF'
}
