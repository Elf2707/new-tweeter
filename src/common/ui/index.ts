export { default as Heart } from './Heart'
export { default as ToggleHeart } from './ToggleHeart'
export { default as FloatedHeartsEffect } from './effects/FloatedHeartsEffect'
export { default as FloatedHeart } from './FloatedHeart'
export { default as HeartButton } from './HeartButton'
export { default as colors } from './colors'
