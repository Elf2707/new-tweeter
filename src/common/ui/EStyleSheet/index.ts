import EStyleSheet from 'react-native-extended-stylesheet'

import colors from '../colors'

EStyleSheet.build({
  ...colors
})
