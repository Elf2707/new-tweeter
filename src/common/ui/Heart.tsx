import React from 'react'
import { View, StyleProp, ViewStyle } from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import colors from './colors'

interface Props {
  style?: StyleProp<ViewStyle>
  filledColor?: string
}

const Heart = ({ filledColor, style }: Props) => {
  const leftStyle = [
    styles.heartHalf,
    styles.leftFilled,
    { backgroundColor: filledColor }
  ]

  const rightStyle = [
    styles.heartHalf,
    styles.rightFilled,
    { backgroundColor: filledColor }
  ]

  return (
    <View style={[styles.container, style]}>
      <View style={leftStyle} />
      <View style={rightStyle} />
    </View>
  )
}

Heart.defaultProps = {
  filledColor: colors.$accentColor
}

export default Heart

const styles = EStyleSheet.create({
  $width: 40,
  $height: 30,
  container: {
    flexDirection: 'row',
    width: '$width',
    height: '$height'
  },
  heartHalf: {
    position: 'absolute',
    width: '$width / 2',
    height: '$height',
    borderTopLeftRadius: '$width / 2',
    borderTopRightRadius: '$width / 2',
    backgroundColor: '$white'
  },
  leftFilled: {
    left: 6,
    backgroundColor: '$accentColor',
    transform: [{ rotate: '-45deg' }]
  },
  rightFilled: {
    right: 6,
    backgroundColor: '$accentColor',
    transform: [{ rotate: '45deg' }]
  },
})
