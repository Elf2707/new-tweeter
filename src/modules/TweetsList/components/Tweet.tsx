import React from 'react'
import { View, Text, Image, GestureResponderEvent } from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import moment from 'moment'

import { HeartButton } from '../../../common/ui'
import { TweetData } from '../model'

interface Props {
  tweet: TweetData
  onLikedPress: (e: GestureResponderEvent, id: string) => void
}

export default function Tweet({ tweet, onLikedPress }: Props) {
  return (
    <View style={styles.container}>
      <View style={styles.infoCont}>
        <Text style={styles.authorText}>{tweet.author}</Text>
        <Text style={styles.dateText}>
          {moment(tweet.createdDate).format('YYYY MMMM DD HH:mm')}
        </Text>

        <Text style={styles.tweetText}>{tweet.about}</Text>

        <HeartButton filled={tweet.liked} onPress={e => onLikedPress(e, tweet.id)} />
      </View>
      <Image style={styles.image} source={{ uri: tweet.imageUrl }} resizeMode="cover" />
    </View>
  )
}

const styles = EStyleSheet.create({
  container: {
    flexDirection: 'row',
    padding: 16
  },
  infoCont: {
    flex: 1,
    paddingRight: 8
  },
  image: {
    width: 100,
    height: 100
  },
  authorText: {
    color: '$textColor',
    fontSize: 20
  },
  tweetText: {
    marginTop: 8,
    marginBottom: 12,
    color: '$subTextColor',
    fontSize: 18
  },
  dateText: {
    color: '$subTextColor',
    fontSize: 12
  }
})
