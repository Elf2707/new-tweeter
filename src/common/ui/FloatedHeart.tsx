import React from 'react'
import { StyleProp, ViewStyle, Animated, StyleSheet } from 'react-native'

import Heart from './Heart'
import colors from './colors'

const NEGATIVE_END_Y = -30
const ANIMATION_END_Y = 30

interface Props {
  style?: StyleProp<ViewStyle>
  onComplete?: () => void
  delay?: number
  forwardPhase?: boolean
  filledColor?: string
  posX: number
  posY: number
}

interface State {
  position: Animated.Value
}

export default class FloatedHeart extends React.Component<Props, State> {
  static defaultPros = {
    onComplete: () => ({}),
    delay: 0,
    forwardPhase: true,
    filledColor: colors.$accentColor
  }

  yAnim: any
  xAnim: any
  opacityAnim: any
  scaleAnim: any
  rotateAnim: any

  constructor(props: Props) {
    super(props)

    const position = new Animated.Value(0)
    this.yAnim = position.interpolate({
      inputRange: [NEGATIVE_END_Y, 0],
      outputRange: [ANIMATION_END_Y, 0]
    })

    this.opacityAnim = this.yAnim.interpolate({
      inputRange: [0, ANIMATION_END_Y / 8, ANIMATION_END_Y],
      outputRange: [0, 1, 0]
    })

    this.scaleAnim = this.yAnim.interpolate({
      inputRange: [0, ANIMATION_END_Y / 2, ANIMATION_END_Y],
      outputRange: [0, 1.1, 1],
      extrapolate: 'clamp'
    })

    this.xAnim = this.yAnim.interpolate({
      inputRange: [0, ANIMATION_END_Y / 2, ANIMATION_END_Y],
      outputRange: [0, props.forwardPhase ? 15 : -15, 0]
    })

    this.rotateAnim = this.yAnim.interpolate({
      inputRange: [
        0,
        ANIMATION_END_Y / 4,
        ANIMATION_END_Y / 3,
        ANIMATION_END_Y / 2,
        ANIMATION_END_Y
      ],
      outputRange: ['0deg', '-2deg', '0deg', '2deg', '0deg']
    })

    this.state = { position }
  }

  componentDidMount() {
    const { onComplete, delay } = this.props
    Animated.timing(this.state.position, {
      delay,
      duration: 1500,
      toValue: NEGATIVE_END_Y,
      useNativeDriver: true
    }).start(onComplete)
  }

  getHeartAnimationStyle(): Partial<ViewStyle> | any {
    return {
      transform: [
        { translateY: this.state.position },
        { translateX: this.xAnim },
        { scale: this.scaleAnim },
        { rotate: this.rotateAnim }
      ],
      opacity: this.opacityAnim
    }
  }

  render() {
    const posStyle = { top: this.props.posY - 100, left: this.props.posX - 16 }

    return (
      <Animated.View
        style={[styles.container, posStyle, this.getHeartAnimationStyle(), this.props.style]}
      >
        <Heart filledColor={this.props.filledColor} />
      </Animated.View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    width: 40,
    height: 30,
    position: 'absolute'
  }
})
