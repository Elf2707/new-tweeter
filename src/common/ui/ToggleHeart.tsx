import React from 'react'
import { View, StyleProp, ViewStyle } from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'

import colors from './colors'
import Heart from './Heart'

interface Props {
  style?: StyleProp<ViewStyle>
  filled?: boolean
  filledColor?: string
}

export default function ToggleHeart({ filled = true, filledColor = colors.$accentColor }: Props) {
  return (
    <View>
      <Heart filledColor={filledColor} />
      {!filled && <Heart style={styles.topHeart} filledColor={colors.$white} />}
    </View>
  )
}

const styles = EStyleSheet.create({
  topHeart: {
    position: 'absolute',
    top: 0,
    left: 0,
    transform: [{ scale: 0.9 }]
  }
})