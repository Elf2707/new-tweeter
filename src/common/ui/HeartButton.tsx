import React from 'react'
import { TouchableWithoutFeedbackProps, TouchableWithoutFeedback, View } from 'react-native'

import ToggleHeart from './ToggleHeart'

interface Props extends TouchableWithoutFeedbackProps {
  filled?: boolean
}

const HeartButton = (props: Props) => {
  return (
    <TouchableWithoutFeedback onPress={props.onPress}>
      <View>
        <ToggleHeart filled={props.filled} />
      </View>
    </TouchableWithoutFeedback>
  )
}

export default HeartButton
