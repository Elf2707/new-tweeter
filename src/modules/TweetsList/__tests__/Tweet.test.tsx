import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Tweet from '../components/Tweet'
import { HeartButton } from '../../../common/ui'

describe('Tweet Tests Suit', () => {
  it('Tweet renders well', () => {
    const testTweet = {
      id: 'id',
      imageUrl: 'image',
      author: 'Author Name',
      about: 'Some tweet about something',
      createdDate: '2018-04-13T05:39:06-02:00',
      liked: false
    }

    const tree = renderer.create(<Tweet tweet={testTweet} onLikedPress={jest.fn()} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Tweet simulate press on liked call onLikedPress with id', () => {
    const testTweet = {
      id: 'id',
      imageUrl: 'image',
      author: 'Author Name',
      about: 'Some tweet about something',
      createdDate: '2018-04-13T05:39:06-02:00',
      liked: false
    }

    const mockFn = jest.fn()
    const wrapper = shallow(<Tweet tweet={testTweet} onLikedPress={mockFn} />)
    const likedBtn = wrapper.find(HeartButton)

    expect(likedBtn).toHaveLength(1)
    likedBtn.simulate('press')

    expect(mockFn).toHaveBeenCalledTimes(1)
    expect(mockFn).toHaveBeenCalledWith(undefined, testTweet.id)
  })
})
