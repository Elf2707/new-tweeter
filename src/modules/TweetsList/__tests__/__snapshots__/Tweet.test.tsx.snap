// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet Tests Suit Tweet renders well 1`] = `
<View>
  <View>
    <Text>
      Author Name
    </Text>
    <Text>
      2018 April 13 10:39
    </Text>
    <Text>
      Some tweet about something
    </Text>
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View>
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                undefined,
                undefined,
                Object {
                  "backgroundColor": "#FF0DFF",
                },
              ]
            }
          />
          <View
            style={
              Array [
                undefined,
                undefined,
                Object {
                  "backgroundColor": "#FF0DFF",
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                undefined,
                undefined,
                Object {
                  "backgroundColor": "#FFF",
                },
              ]
            }
          />
          <View
            style={
              Array [
                undefined,
                undefined,
                Object {
                  "backgroundColor": "#FFF",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "image",
      }
    }
  />
</View>
`;
